Update asset guide - Closes #1104 (#1106)
LiskHQ/lisk-docs: modules/ROOT/pages/guides/app-development/asset.adoc

To give the asset a purpose, it is needed to implement certain logic inside of the asset.

The following sections describe the different components of a transaction asset in detaill.

== The asset class

The asset class always extends from the `BaseAsset`, which is imported from the `lisk-sdk` package.

The properties `name` and `id` are prefilled by the values we used when <<generate-asset,generating the asset skeleton>> in the previous step.

[source,typescript]
----
import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';

export class CreateNFTAsset extends BaseAsset {
  public name = 'createNFT';
  public id = 0;

  // ...
}
----

== The asset schema

The asset schema defines in which format data is sent in the transaction asset.

TIP: For more information about schemas and how they are used in the Lisk SDK, check out the xref:{url_references_schemas}[].

We expect the following data in a transaction, to be able to create a new NFT:

* minPurchaseMargin: The % value of the initial value, that is added to the initial value when purchasing the NFT.
* initValue: The inital value of the NFT.
* name: The name of the NFT.

> Therefore, we adjust the asset schema like this:

Reviewer: ```suggestion
Therefore, the asset schema is adjusted accordingly as shown below:
```

[source,typescript]
----
public schema = {
    $id: "nft/createNFT-asset",
    title: 'CreateNFTAsset transaction asset for nft module',
    type: "object",
    required: ["minPurchaseMargin", "initValue", "name"], <1>
    properties: {
        minPurchaseMargin: {
            dataType: "uint32", <2>
            minimum: 0, <3>
            maximum: 100, <4>
            fieldNumber: 1, <5>
        },
        initValue: {
            dataType: "uint64", <6>
            exclusiveMinimum: 0, <7>
            fieldNumber: 2,
        },
        name: {
            dataType: "string",
            minLength: 3, <8>
            fieldNumber: 3,
        },
    },
};
----

<1> All 3 properties are required to create a new NFT.

> <2> `uin32` is used as data type, because the `minPurchaseMargin` is a small number between 0 and 100.

Reviewer: ```suggestion
<2> `uin32` is used as a data type, because the `minPurchaseMargin` is a minimal number between 0 and 100.
```

<3> The minimum value is set to 0.
<4> The maximum value is set to 100.

> <5> the `fieldNumber` increments by +1 for each property in the transaction asset.

Reviewer: ```suggestion
<5> The `fieldNumber` increments by +1 for each property in the transaction asset.
```

> <6> `uint64` is used as data type, because the `initValue` can have very high values.

Reviewer: ```suggestion
<6> `uint64` is used as a data type, because the `initValue` can have very high values.
```

> <7> The exclusive minimum is set to 0, meaning thee `initValue` needs to be higher than 0 to be valid.

Reviewer: ```suggestion
<7> The exclusive minimum value is set to 0, signifying that the `initValue` needs to be higher than 0 to be valid.
```

<8> The minimum length of the NFT name is set to 3.

== Validating the asset data

The optional function `validate()` validates the data of a transaction asset, before it is passed to the `apply()` function.

> If one of these conditions is not fulfilled, the transaction won't be processed, and an error should be thrown.

Reviewer: ```suggestion
If one of these conditions is not fulfilled, then the transaction will not be processed, and an error should be thrown.
```

[source,typescript]
----
public validate({ asset }: ValidateAssetContext<{}>): void {
    if (asset.name === "reservedName") {
      throw new Error("Illegal NFT name");
    }
}
----

> If the validation doesn't throw any errors, it means the validations have been successful, and the data will be applied as defined in the `apply()` function.

Reviewer: ```suggestion
If the validation does not throw any errors, that signifies they been successful and the data will be applied as defined in the `apply()` function.
```

== Defining the asset logic

The most important part of the transaction asset is the `apply()` function.
It contains the logic of how the data in the transaction asset should be applied on the blockchain.

In this example, we use the transaction data to create a new NFT token, which is added to the senders account.

> To get and set the blockchain state, the `stateStore` is used again, which we know already from the xref:{url_guides_module_statestore}[lifecycle hooks] of the module guide.

Reviewer: ```suggestion
To get and set the blockchain state, the `stateStore` is used again, which we are already aware of from the xref:{url_guides_module_statestore}[lifecycle hooks] in the module guide.
```
